llvm/unittests/IR/ConstantFPRangeTest.cpp

```diff
@@ -0,0 +1,370 @@
+  EXPECT_FALSE(PosZero.contains(APFloat::getZero(Sem, /*Negative=*/true)));
+  EXPECT_TRUE(NegZero.contains(APFloat::getZero(Sem, /*Negative=*/true)));
+  EXPECT_FALSE(NegZero.contains(APFloat::getZero(Sem, /*Negative=*/false)));
+  EXPECT_TRUE(Zero.contains(PosZero));
+  EXPECT_TRUE(Zero.contains(NegZero));
+  EXPECT_TRUE(Denormal.contains(APFloat::getSmallest(Sem)));
+  EXPECT_FALSE(Denormal.contains(APFloat::getSmallestNormalized(Sem)));
+  EXPECT_TRUE(PosInf.contains(APFloat::getInf(Sem, /*Negative=*/false)));
+  EXPECT_TRUE(NegInf.contains(APFloat::getInf(Sem, /*Negative=*/true)));
+  EXPECT_TRUE(NaN.contains(APFloat::getQNaN(Sem)));
+  EXPECT_TRUE(NaN.contains(APFloat::getSNaN(Sem)));
+  EXPECT_TRUE(NaN.contains(SNaN));
+  EXPECT_TRUE(NaN.contains(QNaN));
+
+  EXPECT_TRUE(Some.contains(APFloat(3.0)));
+  EXPECT_TRUE(Some.contains(APFloat(-3.0)));
+  EXPECT_FALSE(Some.contains(APFloat(4.0)));
+  APFloat Next1(3.0);
+  Next1.next(/*nextDown=*/true);
+  EXPECT_TRUE(Some.contains(Next1));
+  APFloat Next2(3.0);
+  Next2.next(/*nextDown=*/false);
+  EXPECT_FALSE(Some.contains(Next2));
+  EXPECT_TRUE(Some.contains(Zero));
+  EXPECT_TRUE(Some.contains(Some));
+  EXPECT_TRUE(Some.contains(One));
+  EXPECT_FALSE(Some.contains(NaN));
+  EXPECT_FALSE(Some.contains(PosInf));
+  EXPECT_TRUE(SomePos.contains(APFloat(3.0)));
+  EXPECT_FALSE(SomeNeg.contains(APFloat(3.0)));
+  EXPECT_TRUE(SomeNeg.contains(APFloat(-3.0)));
+  EXPECT_FALSE(SomePos.contains(APFloat(-3.0)));
+  EXPECT_TRUE(Some.contains(SomePos));
+  EXPECT_TRUE(Some.contains(SomeNeg));
+}
+
+TEST_F(ConstantFPRangeTest, Equality) {
+  EXPECT_EQ(Full, Full);
+  EXPECT_EQ(Empty, Empty);
+  EXPECT_EQ(One, One);
+  EXPECT_EQ(Some, Some);
+  EXPECT_NE(Full, Empty);
+  EXPECT_NE(Zero, PosZero);
+  EXPECT_NE(One, NaN);
+  EXPECT_NE(Some, One);
+  EXPECT_NE(SNaN, QNaN);
+}
+
+TEST_F(ConstantFPRangeTest, SingleElement) {
+  EXPECT_EQ(Full.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(Empty.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(Finite.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(Zero.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(NaN.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(SNaN.getSingleElement(), static_cast<APFloat *>(nullptr));
+  EXPECT_EQ(QNaN.getSingleElement(), static_cast<APFloat *>(nullptr));
+
+  EXPECT_EQ(*One.getSingleElement(), APFloat(1.0));
+  EXPECT_EQ(*PosZero.getSingleElement(), APFloat::getZero(Sem));
+  EXPECT_EQ(*PosInf.getSingleElement(), APFloat::getInf(Sem));
+
+  EXPECT_FALSE(Full.isSingleElement());
+  EXPECT_FALSE(Empty.isSingleElement());
+  EXPECT_TRUE(One.isSingleElement());
+  EXPECT_FALSE(Some.isSingleElement());
+  EXPECT_FALSE(Zero.isSingleElement());
+}
+
+TEST_F(ConstantFPRangeTest, Enumerate) {
+  constexpr unsigned NNaNValues = (1 << 8) - 2 * ((1 << 3) - 1);
+  constexpr unsigned Expected = 4 * ((NNaNValues + 1) * NNaNValues / 2 + 1);
```

Comment:
We have 117616 ranges here. It is impossible to exhaustively test binary operations.

Reply:
Hm yeah, I didn't think the numbers here through. Probably the best we can do instead is to have a list of interesting values (-inf, -largest_normal, -smallest_normal, -largest_subnormal, -smallest_subnormal, -0, +0, ...) plus a few normal numbers to use as boundary values. Not as good as full exhaustive coverage, but at least it makes sure that all the special cases are treated correctly.

Reply:
Should test with a bit in every mantissa position

Reply:
Reduced to 2664 ranges.